test/model/internal_link_test.dart

```diff
@@ -0,0 +1,486 @@
+      check(decodeHashComponent('some_text')).equals('some_text');
+      check(decodeHashComponent('some.20text')).equals('some text');
+      check(decodeHashComponent('some.2Etext')).equals('some.text');
+
+      check(decodeHashComponent('na.C3.AFvet.C3.A9')).equals('naïveté');
+      check(decodeHashComponent('.C2.AF.5C_(.E3.83.84)_.2F.C2.AF')).equals('¯\\_(ツ)_/¯');
+
+      // malformed dot-encoding
+      check(decodeHashComponent('some.text')).isNull();
+      check(decodeHashComponent('some.2gtext')).isNull();
+      check(decodeHashComponent('some.arbitrary_text')).isNull();
+
+      // malformed UTF-8
+      check(decodeHashComponent('.88.99.AA.BB')).isNull();
+    });
+
+    test('parse correctly in stream and topic operands', () {
+      final streams = [
+        eg.stream(streamId: 1, name: 'some_stream'),
+        eg.stream(streamId: 2, name: 'some stream'),
+        eg.stream(streamId: 3, name: 'some.stream'),
+      ];
+      final testCases = [
+        ('/#narrow/stream/some_stream',
+          const StreamNarrow(1)),
+        ('/#narrow/stream/some.20stream',
+          const StreamNarrow(2)),
+        ('/#narrow/stream/some.2Estream',
+          const StreamNarrow(3)),
+        ('/#narrow/stream/some_stream/topic/some_topic',
+          const TopicNarrow(1, 'some_topic')),
+        ('/#narrow/stream/some_stream/topic/some.20topic',
+          const TopicNarrow(1, 'some topic')),
+        ('/#narrow/stream/some_stream/topic/some.2Etopic',
+          const TopicNarrow(1, 'some.topic')),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+  });
+
+  group('Stream link parsing edge cases', () {
+    final streams = [
+      eg.stream(streamId: 1, name: "general"),
+    ];
+
+    test('basic', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/1-general', const StreamNarrow(1)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
+
+    test('if stream not found, use stream ID anyway', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/123-topic', const StreamNarrow(123)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
+
+    test('on stream link with wrong name, ID wins', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/1-nonsense', const StreamNarrow(1)),
+        ('#narrow/stream/1-', const StreamNarrow(1)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
+
+    test('on malformed stream link: reject', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/-1', null),
+        ('#narrow/stream/1nonsense-general', null),
+        ('#narrow/stream/-general', null),
```

Comment:
(Here it's OK that these are all in one `test` call, because that's not a regression from the zulip-mobile tests.)